lib/testpool/single-process-test-pool.js

```diff
@@ -0,0 +1,135 @@
+const fs = require('fs');
+const Bluebird = require('bluebird');
+const debug = require('debug')('ava:pool');
+const {failTypes} = require('../fork');
+const precompilerHook = require('../worker/precompiler-hook');
+const TestPool = require('./test-pool');
+const isolatedRequire = require('./isolated-require');
+
+const preTestPath = require.resolve('./test-pre-require.js');
+const postTestPath = require.resolve('./test-post-require.js');
+
+const preTestContent = fs.readFileSync(preTestPath, 'utf8').toString();
+const postTestContent = fs.readFileSync(postTestPath, 'utf8').toString();
+
+// Keep track of callbacks so we only need to register
+// one event listener for each event type
+const uncaughtListeners = new Set();
+const unhandledListeners = new Set();
+
+process.on('uncaughtException', err => {
+	uncaughtListeners.forEach(cb => cb(err));
+});
+
+process.on('unhandledRejection', (reason, promise) => {
+	unhandledListeners.forEach(cb => cb(reason, promise));
+});
+
+// We wait to clean up rejection/exception listeners
+// until we are finishing up since an exception could
+// be delayed
+const allListenerCleanUps = new Set();
+const doCleanUp = () => {
+	allListenerCleanUps.forEach(cleanUp => cleanUp());
+};
+
+// All tests are run in the same process
+class SingleProcessTestPool extends TestPool {
+	constructor(files, options) {
+		super(files, options);
+		this.resolves = new Set();
+	}
+
+	bail() {
+		debug('bailing single process');
+		this.resolves.forEach(resolve => resolve());
+	}
+
+	run() {
+		const {options, runOnlyExclusive} = this;
+		const concurrency = options.serial ? 1 : 2;
```

Comment:
I'm not sure we need to have any test file concurrency within the process. That would also simplify the implementation.